C compilation error: invalid operand types for binary operation

Repo: https://github.com/leNicDev/chip8-v

<!-- It is advisable to update all relevant modules using `v outdated` and `v install` -->

<details>
<summary>V version: V 0.4.9 2aa5651, press to see full `v doctor` output</summary>

|V full version      |V 0.4.9 d313355.2aa5651
|:-------------------|:-------------------
|OS                  |linux, "CachyOS"
|Processor           |32 cpus, 64bit, little endian, AMD Ryzen 9 9950X 16-Core Processor
|Memory              |30.15GB/46.17GB
|                    |
|V executable        |/home/nic/v/v
|V last modified time|2025-01-06 09:27:48
|                    |
|V home dir          |OK, value: /home/nic/v
|VMODULES            |OK, value: /home/nic/.vmodules
|VTMP                |OK, value: /tmp/v_1000
|Current working dir |OK, value: /home/nic/dev/chip8-v/src
|                    |
|Git version         |git version 2.47.1
|V git status        |Error: fatal: not a git repository (or any parent up to mount point /)
Stopping at filesystem boundary (GIT_DISCOVERY_ACROSS_FILESYSTEM not set).
|.git/config present |false
|                    |
|cc version          |cc (GCC) 14.2.1 20240910
|gcc version         |gcc (GCC) 14.2.1 20240910
|clang version       |clang version 18.1.8
|tcc version         |tcc version 0.9.28rc 2024-07-31 HEAD@1cee0908 (x86_64 Linux)
|tcc git status      |N/A
|emcc version        |N/A
|glibc version       |ldd (GNU libc) 2.40

</details>

**What did you do?**
`./v -g -o vdbg cmd/v && ./vdbg  main.v && main`
```v
module main

import chip8 { System }
import gg
import gx

fn main() {
	mut system := System{}
	system.reset()

	ui_thread := spawn create_ui(&system)

	system.run()

	ui_thread.wait()
}

fn create_ui(emulator &System) {
	mut context := gg.new_context(
		bg_color:     gx.rgb(0, 0, 0)
		width:        640
		height:       320
		window_title: 'CHIP-8 Emulator'
		frame_fn:     frame
	)
	context.user_data = emulator
	context.run()
}

fn frame(mut ctx gg.Context) {
	mut system := &System{}

	unsafe {
		system = &System(ctx.user_data)
	}
	image := ctx.new_streaming_image(64, 32, 4, gg.StreamingImageConfig{
		pixel_format: .rgba8
		mag_filter:   .nearest
	})
	rlock system.gfx {
		rgba := bool_to_rgba8(system.gfx)
		ctx.update_pixel_data(image, &rgba[0])
	}

	ctx.draw_image(0, 0, 64, 32, ctx.get_cached_image_by_idx(image))
}

fn bool_to_rgba8(bw [64 * 32]bool) []u8 {
	mut rgba := []u8{len: bw.len * 4}

	for i, is_white in bw {
		idx := i * 4
		if is_white {
			rgba[idx] = 255
			rgba[idx + 1] = 255
			rgba[idx + 2] = 255
			rgba[idx + 3] = 255
		} else {
			rgba[idx] = 0
			rgba[idx + 1] = 0
			rgba[idx + 2] = 0
			rgba[idx + 3] = 255
		}
	}

	return rgba
}

```

**What did you see?**
```
================== C compilation error (from tcc): ==============
cc: /tmp/v_1000/main.01JH39ERZEWPV67G0NSZ18HP3Q.tmp.c:15606: error: invalid operand types for binary operation
=================================================================
(You can pass `-cg`, or `-show-c-output` as well, to print all the C error messages).
builder error: 
==================
C error found. It should never happen, when compiling pure V code.
This is a V compiler bug, please report it using `v bug file.v`,
or goto https://github.com/vlang/v/issues/new/choose .
You can also use #help on Discord: https://discord.gg/vlang .
```

**What did you expect to see?**

Successful compilation


> [!NOTE]
> You can use the 👍 reaction to increase the issue's priority for developers.
>
> Please note that only the 👍 reaction to the issue itself counts as a vote.
> Other reactions and those to comments will not be taken into account.

--- a/vlib/v/gen/c/index.v
+++ b/vlib/v/gen/c/index.v
@@ -372,6 +372,9 @@ fn (mut g Gen) index_of_fixed_array(node ast.IndexExpr, sym ast.TypeSymbol) {
 		} else {
 			g.expr(node.left)
 		}
+		if node.left_type.has_flag(.shared_f) {
+			g.write('.val')
+		}
 	}
 	g.write('[')
 	if g.is_direct_array_access || g.pref.translated || node.index is ast.IntegerLiteral {

--- a/vlib/v/tests/concurrency/shared_array_indexing_test.v
+++ b/vlib/v/tests/concurrency/shared_array_indexing_test.v
@@ -1,0 +1,18 @@
+struct Foo {
+mut:
+	bar shared [10]bool
+}
+
+fn test_main() {
+	mut a := Foo{
+		bar: [10]bool{}
+	}
+	lock a.bar {
+		a.bar[0] = true
+		a.bar[1] = false
+	}
+	rlock a.bar {
+		assert a.bar[0] == true
+		assert a.bar[1] == false
+	}
+}

